### ISSUE TITLE
admin: /reverse_proxy/upstreams endpoint should provide info on active health checks

### ISSUE THREAD

**Issue by ottenhoff** [caddyserver/caddy#6135]
Use case: I know how to health-check my upstreams separately, but I want to quickly understand which upstreams *Caddy* believes are healthy or unhealthy. 

I am using active health checks

```
$ curl http://localhost:2019/reverse_proxy/upstreams | jq
[
  {
    "address": "10.1.67.25:8156",
    "num_requests": 0,
    "fails": 0
  },
  {
    "address": "10.2.67.25:9286",
    "num_requests": 0,
    "fails": 0
  }
]
```

Other load balancers provide info on each upstream:

1. number of active checks run since last restart
2. number of active checks that failed
3. Healthy/unhealthy flag
4. "Last" check status (e.g., if you require 3 successive checks to become healthy then the last check would be an early indicator that your upstream is on its way to healthy)

**Comment by francislavoie** [comment]
The problem is, health status is not actually in the upstream storage. Health is a factor of the proxy config compared to the storage. See https://caddyserver.com/docs/api#get-reverse-proxyupstreams which explains.

Basically, you could have more than one proxy configured, each with different settings for [`max_fails`](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#max_fails) so it might healthy for one proxy but not the other.

**Comment by ottenhoff** [comment]
Right, for a passive check, I understand that providing the health of an upstream depends on the proxy config. What about an active check? I'd like to know how many active health checks are passing/failing

**Comment by francislavoie** [comment]
The active health checks are stored on the individual proxy's references to the upstream, so it's not in the global store. So we don't have any way to access it from the admin API at this time.

Basically, the current data structures make it impossible to track this kind of information from the admin API. We'll need to rework things at some point to make it possible.

What you _can_ do though is hook into [events](https://caddyserver.com/docs/caddyfile/options#event-options) which would let you do something in reaction to a change in health status of an upstream.

**Comment by mholt** [comment]
Yeah, passive "health status" (up/down) is a matter of interpreting the number of fails currently remembered. It's up to you to decide whether that constitutes down or not. You could simply use the same parameters as your config, at least for now.

I have some ideas on how to make this more usable with API endpoints but would probably need a sponsorship to fund its development at this point with a lot on my plate right now.

**Comment by ottenhoff** [comment]
> The active health checks are stored on the individual proxy instance's references to the upstream, so it's not in the global store. So we don't have any way to access it from the admin API at this time.

> Basically, the current data structures make it impossible to track this kind of information from the admin API. We'll need to rework things at some point to make it possible.

Is this still accurate in 2025?

**Comment by mholt** [comment]
Yeah. We just need to think carefully about how to expose that information to the API. It's a bit tricky.

**Comment by mildred** [comment]
We are using Caddy with reverse_proxy, and we have a script that does rolling restart of the servers behind Caddy. In order to perform a rolling restart we have the following procedure:

- bring a new instance and wait for it to be ready
- add the new instance to the upstreams using the Caddy config API
- tell the old instance to become unhealthy
- wait a bit until Caddy notices
- remove the old instance from the upstreams using the Caddy Config API
- shut down the old instance

We would like to avoid unnecessary wait while shutting down an instance. Detecting when the health check becomes unhealthy would be great to allow that.
